chia/_tests/core/mempool/test_mempool_manager.py

```diff
@@ -1937,3 +1939,74 @@ async def get_coin_records(coin_ids: Collection[bytes32]) -> List[CoinRecord]:
+            ConditionOpcode.CREATE_COIN,
+            ConditionOpcode.ASSERT_HEIGHT_ABSOLUTE,
+            ConditionOpcode.ASSERT_SECONDS_ABSOLUTE,
+            ConditionOpcode.CREATE_COIN_ANNOUNCEMENT,
+            ConditionOpcode.REMARK,
+        ]
+        puzzles_and_coins = []
+        conditions = []
+        for i in range(15):
+            random_conditions_count = random.randint(1, len(conditions_pool))
+            opcodes = random.sample(conditions_pool, random_conditions_count)
+            for opcode in opcodes:
+                if opcode == ConditionOpcode.RESERVE_FEE:
+                    conditions.append([opcode, uint64(i)])
+                elif opcode == ConditionOpcode.CREATE_COIN:
+                    conditions.append([opcode, IDENTITY_PUZZLE_HASH, uint64(i)])
+                elif opcode == ConditionOpcode.ASSERT_HEIGHT_ABSOLUTE:
+                    conditions.append([opcode, uint32(i)])
+                elif opcode == ConditionOpcode.ASSERT_SECONDS_ABSOLUTE:
+                    conditions.append([opcode, uint64(i)])
+                elif opcode == ConditionOpcode.CREATE_COIN_ANNOUNCEMENT:
+                    conditions.append([opcode, i.to_bytes(32, "big")])
+                elif opcode == ConditionOpcode.REMARK:
+                    conditions.append([opcode, i])
+            # These puzzles are of the form (q . conditions)
+            puzzle = SerializedProgram.to((1, conditions))
+            ph = puzzle.get_tree_hash()
+            for _ in range(2):
+                await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
+            coin_records = await full_node_api.full_node.coin_store.get_coin_records_by_puzzle_hash(False, ph)
+            puzzles_and_coins.append((puzzle, ph, coin_records[0].coin))
+        # Use the farmed coins to fill the mempool with bundles that spend them
+        for puzzle, ph, coin in puzzles_and_coins:
+            coin_spend = make_spend(coin, puzzle, SerializedProgram.to([]))
+            sb = SpendBundle([coin_spend], G2Element())
+            await send_to_mempool(full_node_api, sb)
+
+    [[full_node_api], _, _] = simulator_and_wallet
+    TEST_MAX_BLOCK_CLVM_COST = uint64(150_000_000)
+    assert full_node_api.full_node._mempool_manager is not None
+    full_node_api.full_node._mempool_manager.mempool.mempool_info = dataclasses.replace(
+        full_node_api.full_node._mempool_manager.mempool.mempool_info,
+        max_block_clvm_cost=CLVMCost(TEST_MAX_BLOCK_CLVM_COST),
+    )
+    full_node_api.full_node.blockchain.constants = full_node_api.full_node.blockchain.constants.replace(
+        MAX_BLOCK_COST_CLVM=TEST_MAX_BLOCK_CLVM_COST
+    )
+    # Generate blocks with various combinations of puzzles
+    for _ in range(10):
+        await fill_mempool_with_test_sbs(full_node_api)
+        # Farm the block to make sure we're passing block validation
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(IDENTITY_PUZZLE_HASH))
```

Comment:
shouldn't there be some check at the end? to ensure the final block ended up not being empty for instance?
We have a safety net in the farmer that if it ends up with an invalid block, it farms an empty block instead

Reply:
Yep, fair enough. Done.